wgpu-core/src/indirect_validation.rs

```diff
@@ -0,0 +1,429 @@
+use std::sync::atomic::AtomicBool;
+
+use thiserror::Error;
+
+use crate::{
+    device::DeviceError,
+    pipeline::{CreateComputePipelineError, CreateShaderModuleError},
+};
+
+#[derive(Clone, Debug, Error)]
+#[non_exhaustive]
+pub enum CreateDispatchIndirectValidationPipelineError {
+    #[error(transparent)]
+    DeviceError(#[from] DeviceError),
+    #[error(transparent)]
+    ShaderModule(#[from] CreateShaderModuleError),
+    #[error(transparent)]
+    ComputePipeline(#[from] CreateComputePipelineError),
+}
+
+/// This machinery requires the following limits:
+///
+/// - max_bind_groups: 2,
+/// - max_dynamic_storage_buffers_per_pipeline_layout: 1,
+/// - max_storage_buffers_per_shader_stage: 2,
+/// - max_storage_buffer_binding_size: 3 * min_storage_buffer_offset_alignment,
+/// - max_push_constant_size: 4,
+/// - max_compute_invocations_per_workgroup 1
+///
+/// These are all indirectly satisfied by `DownlevelFlags::INDIRECT_EXECUTION`, which is also
+/// required for this module's functionality to work.
+#[derive(Debug)]
+pub struct IndirectValidation {
+    module: Box<dyn hal::DynShaderModule>,
+    dst_bind_group_layout: Box<dyn hal::DynBindGroupLayout>,
+    src_bind_group_layout: Box<dyn hal::DynBindGroupLayout>,
+    pipeline_layout: Box<dyn hal::DynPipelineLayout>,
+    pipeline: Box<dyn hal::DynComputePipeline>,
+    dst_buffer_0: Box<dyn hal::DynBuffer>,
+    dst_buffer_1: Box<dyn hal::DynBuffer>,
+    dst_bind_group_0: Box<dyn hal::DynBindGroup>,
+    dst_bind_group_1: Box<dyn hal::DynBindGroup>,
+    is_next_dst_0: AtomicBool,
```

Comment:
issue: @cwfitzgerald was concerned (and now I am, too) that this usage of `AtomicBool` is not correct for multi-threaded usage. Encoding of commands can happen on multiple different threads, and the swapping done here is greatly complicated by the fact that the commands that necessitate swapping may not be submitted in the order this boolean attempts to enforce.

Reply:
Yeah, implementing this optimization sounds more complicated than I thought. I think it would be a problem even in non multithreaded use-cases because even on a single thread users can create multiple command buffers that then get submitted in a different order (than the order they have been created in).

I will drop the commit and create a new issue to explore this further.